### ISSUE TITLE
Add support for caching-sha2-password

### ISSUE THREAD

**Issue by kzemek** [mysql-otp/mysql-otp#83]
`caching-sha2-password` [became the default authentication plugin in MySQL 8.0.4](https://dev.mysql.com/doc/refman/8.0/en/upgrading-from-previous-series.html#upgrade-caching-sha2-password)

Trying to connect to a server with new default, the connection process crashes:

```
2018-04-24 13:43:18.523 [error] <0.2253.0> CRASH REPORT Process <0.2253.0> with 0 neighbours exited with reason: {auth_method,<<"caching_sha2_password">>} in mysql_protocol:build_handshake_response/5 line 310 in gen_server:init_it/6 line 352
2018-04-24 13:43:18.529 [error] <0.2250.0>@mongoose_rdbms:connect:450 Database connection attempt with {mysql,"localhost","ejabberd","ejabberd","mongooseim_secret",[{verify,verify_peer},{cacertfile,"priv/ssl/cacert.pem"}]} resulted in {error,{{auth_method,<<"caching_sha2_password">>},[{mysql_protocol,build_handshake_response,5,[{file,"/home/travis/build/esl/MongooseIM/_build/default/lib/mysql/src/mysql_protocol.erl"},{line,310}]},{mysql_protocol,handshake,7,[{file,"/home/travis/build/esl/MongooseIM/_build/default/lib/mysql/src/mysql_protocol.erl"},{line,64}]},{mysql,init,1,[{file,"/home/travis/build/esl/MongooseIM/_build/default/lib/mysql/src/mysql.erl"},{line,499}]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,328}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}}. Retrying in 2 seconds.
```

**Comment by shantiii** [comment]
Just a heads up that I'm working on this now - the [flow for this auth](https://dev.mysql.com/doc/dev/mysql-server/latest/page_caching_sha2_authentication_exchanges.html) is well-documented, at least. Hope to have a PR up by Monday.

**Comment by shantiii** [comment]
Looks like this was a little more involved than I had hoped - the current handshake code has a strong assumption that the authentication call and response will take only 1 roundtrip, when the caching_sha2_password flow can include a variable number of roundtrips. This is going to involve a little bit of restructuring of handshake, but should be ok in the end.

**Comment by juhlig** [comment]
@zuiderkwast I'm currently trying my hand at this, but turns out this is not exactly trivial. Got a prototype mostly working at least, but all that special treatment in various places is cluttering up the `mysql_protocol` module. I'll try to think something up to factor out authentication into separate modules, that would maybe even allow for easier integration of other auth methods in the future, plugin-like. But no promises ;)

On a side note, I have been experimenting with travis and MySQL 8 for this to be testable. I tried [this approach](https://travis-ci.community/t/how-can-i-use-mysql-8-0-on-xenial-distribution-currently-it-says-5-7/2728), seems to work to some extend, but many of the tests for `mysql-otp` fail for one reason or another. Anyway, please take a look, maybe you can make something out of that. I'm not too familiar with travis ;)

**Comment by zuiderkwast** [comment]
Solving the Travis issue first sounds like the right approach. The example you found with xenial looks promising, but I don't have much Travis experience either. It should be straight-forward to try it out in a fork, I assume.

Regarding the implementation, do you base the work on #101? I don't think it makes the logic too cluttered.

Premature refactoring can sometimes cause problems, if it turns out the plugin interface needs to be changed whenever the next auth plugin is implemented, so I wouldn't worry too much about that for the first auth plugin. If however there are more plugins to be added in the future, factoring them out into separate modules would definitely be the way to go. MariaDB has some auth methods, notably ed25519 (see [blog post](https://mariadb.org/history-of-mysql-mariadb-authentication-protocols/)).

**Comment by juhlig** [comment]
> Solving the Travis issue first sounds like the right approach. The example you found with xenial looks promising, but I don't have much Travis experience either. It should be straight-forward to try it out in a fork, I assume.

I think this can only be solved "in tandem", ie both at once. As I said, many of the current tests don't pass with MySQL 8, even when it is tweaked to accept `mysql_native_password` authentication. Notably, tests that try to connect with a non-existing user fail because the server does not reply with "... Access denied..." but with "Client does not support authentication protocol requested by server; consider upgrading MySQL client" (strangely, however, they sometimes pass, don't know why). There are others that fail for other reasons which I didn't look into.

FWIW, [this](https://github.com/juhlig/mysql-otp/blob/travis_mysql8/.travis.yml) is the `.travis.yml` I got it working with (reduced the OTP releases to only 22 for quicker testing), it creates jobs for both MySQL 5.7 (what happens to come with Xenial) and 8 (with both being widely used and considering the set of differences between the two, I think we should test with both, not only 8). I guess using environment variables are not really the way to do it, but as I said, I'm not very familiar with travis, and it does the job of using MySQL 8 ;)

> Regarding the implementation, do you base the work on #101? I don't think it makes the logic too cluttered.

I'm using it as a kind of reference/inspiration/check, but I'm doing a more or less new implementation.

> Premature refactoring can sometimes cause problems, if it turns out the plugin interface needs to be changed whenever the next auth plugin is implemented, so I wouldn't worry too much about that for the first auth plugin. If however there are more plugins to be added in the future, factoring them out into separate modules would definitely be the way to go. MariaDB has some auth methods, notably ed25519 (see [blog post](https://mariadb.org/history-of-mysql-mariadb-authentication-protocols/)).

Ok, with `mysql_native_password` and `caching_sha2_password` being built-ins in the server, it can just as well be built-in in the client, I agree. I'll keep the plugin auth idea in the back of my head, though, because one of the ideas behind pluggable authentication is that users can make own custom auth plugins for their server instances and want the same for their clients. Granted, I guess nobody in his right mind would _want_ to develop custom auth plugins, but the ed25519 plugin in MariaDB is an example where it could be useful.

**Comment by zuiderkwast** [comment]
Implemented in #152.
